lib/poparser.js

```diff
@@ -3,53 +3,95 @@ const sharedFuncs = require('./shared');
+    defaultCharset,
+    validation
+  } = normalizeParseArgs(arg0, arg1);
+
+  const parser = new Parser(buffer, defaultCharset, validation);
 
   return parser.parse();
 };
 
 /**
- * Parses a PO stream, emits translation table in object mode
-
- * @param defaultCharset Default charset to use
- * @param options Stream options
- * @param validation Enables validation of parsed content, if set to true
+ * Normalizes arguments passed into `stream` function
+ *
+ * @typedef {{ defaultCharset?: string, validation?: boolean }} Options
+ * @typedef {{ transformOptions: import('readable-stream').TransformOptions, defaultCharset?: string, validation?: boolean }} ReturnType
  * @type {{
- *    () => Stream;
- *    (defaultCharset: string) => Stream;
- *    (defaultCharset: string, options: import('readable-stream').TransformOptions) => Stream;
- *    (defaultCharset: string, options: import('readable-stream').TransformOptions, validation: boolean) => Stream;
- *    (defaultCharset: string, validation: boolean) => Stream;
- *    (options: import('readable-stream').TransformOptions) => Stream;
- *    (options: import('readable-stream').TransformOptions, validation: boolean) => Stream;
- *    (validation: boolean) => Stream;
+ *    () => ReturnType;
+ *    (defaultCharset: string) => ReturnType;
+ *    (defaultCharset: string, transformOptions: import('readable-stream').TransformOptions) => ReturnType;
+ *    (transformOptions: import('readable-stream').TransformOptions) => ReturnType;
+ *    (options: Options) => ReturnType;
+ *    (options: Options, transformOptions: import('readable-stream').TransformOptions) => ReturnType;
  * }}
  */
-module.exports.stream = function (defaultCharset, options, validation) {
-  const args = sharedFuncs.unshiftOptionalArgs(
-    ['string', 'object', 'boolean'],
-    [defaultCharset, options, validation]
-  );
+function normalizeStreamArgs (arg0 = {}, arg1 = {}) {
+  return typeof arg0 === 'string'
+    ? { defaultCharset: arg0, transformOptions: arg1 }
+    : 'defaultCharset' in arg0 || 'validation' in arg0
+      ? { defaultCharset: arg0.defaultCharset, validation: arg0.validation, transformOptions: arg1 }
+      : { transformOptions: arg0 };
+}
 
-  return new PoParserTransform(...args);
+/**
+ * Parses a PO stream, emits translation table in object mode
+ *
+ * @typedef {{ defaultCharset: strubg, validation: boolean }} Options
+ * @param defaultCharset Default charset to use
+ * @param transformOptions Stream options
+ * @param options Optional options with defaultCharset and validation
+ * @type {{
+*    () => Stream;
+*    (defaultCharset: string) => Stream;
+*    (defaultCharset: string, transformOptions: import('readable-stream').TransformOptions) => Stream;
+*    (transformOptions: import('readable-stream').TransformOptions) => Stream;
+*    (options: Options) => Stream;
+*    (options: Options, transformOptions: import('readable-stream').TransformOptions) => Stream;
+* }}
+*/
+function stream (arg0, arg1) {
```

Comment:
First of all, I think only 2 signatures should be 'promoted':
* current usage: defaultCharset, transformOptions
* new usage: options, transformOptions
`transformOptions` should never be the first argument.

As the new signature looks way better (thinking about future options), let's write it like that:
```javascript
function stream (options, transformOptions) { 
```
The code then becomes easier to reason about (`arg0` and `arg1` don't have any meaning).

I'll leave a comment below (with `PoParserTransform`) to clarify more.

Reply:
> `transformOptions` should never be the first argument.

If I am reading these [lines of code](https://github.com/smhg/gettext-parser/blob/master/lib/poparser.js#L479-L482) correctly, I am afraid having `transformOptions` as the first argument is already permitted. Correct me if I'm wrong but my understanding is that we would like the function signature to stay fully backward compatible. That is the reason for using `arg0` and `arg1`.

I think that if this change is planned to be released with library's major version bump then it would make sense to introduce breaking changes, to forbid having `transformOptions` as the first argument and only support those function signatures:

```js
() => Stream;
(defaultCharset: string) => Stream;
(defaultCharset: string, transformOptions: import('readable-stream').TransformOptions) => Stream;
(options: Options) => Stream;
(options: Options, transformOptions: import('readable-stream').TransformOptions) => Stream;
```

Please let me know if that is the case.

Reply:
@smhg In accordance with https://github.com/smhg/gettext-parser/pull/76#issuecomment-1478410116 I made necessary [adjustments](https://github.com/smhg/gettext-parser/pull/76/commits/caa544530d6c04cf1e0fddebae0a18f2a88bfa51).